egressmap: add support for multiple gateway nodes per policy

jibi opened the pull request:
This commit lays the groundwork (mostly from a datapath perspective) to
implement high availability egress grateway by adding support for
multiple gateway nodes per policy, sharing all the same egress IP.

The format of the egress policy map has been updated to allow storing
multiple gateways per policy, and a new CT map has been introduced to
pin connections to a specific gateway.

The CLI logic has also been extended to support managing multiple
gateways for a single policy.

Note that these changes assume the router in front of the gateways will
keep track of which ECMP route a connection belongs to (in order to
forward reply traffic back to the correct gateway).
This in practice is not always true, and in a subsequent PR a more
general approach will be introduced to make sure reply traffic will
eventually hit the correct gateway, regarless of which one the router
forwards the traffic to.

Lastly, for the case of an egress policy with a single gateway, this
change should be backward compatible with the current egress gateway
implementation.

pchaigno commented on bpf/custom/bpf_custom.c:
```diff
@@ -4,6 +4,8 @@
 #include <bpf/ctx/skb.h>
 #include <bpf/api.h>
 
+#include "node_config.h"
```
Why is this change required?

jibi replied:
I initially added it to fix a test (I don't remember anymore if it was `make -C test/bpf` or `make -C bpf go_prog_test`) but now it seems both tests are compiling fine even without it, so I'll just remove it (I may have accidentally fixed this case with some other change)

pchaigno commented on daemon/cmd/datapath.go:
```diff
@@ -438,6 +434,12 @@ func (d *Daemon) initMaps() error {
+		if err := egressmap.InitEgressMaps(); err != nil {
+			return err
+		}
+	}
```
Did you move this check just to be with the other conditional map inits or was there another reason? (I.e., is there a dependency I'm missing on the order?)

jibi replied:
Just to be in line with the other conditional map inits

pchaigno replied:
Ok :+1:

pchaigno commented on bpf/lib/common.h:
```diff
@@ -245,9 +249,20 @@ struct endpoint_info {
+	struct bpf_lpm_trie_key lpm_key;
+	__be32 saddr;
+	__be32 daddr;
+};
```
This changes the size of the key and therefore will need consideration on upgrades & downgrades to not flush the map, no?

jibi replied:
Shouldn't be an issue as I renamed the old `cilium_egress_v4` to `cilium_egress_policy_v4` (in order to have consistent naming with the other related map, `cilium_egress_ct_v4`)

pchaigno replied:
Renaming also typically requires special consideration for upgrades/downgrades. We currently don't wait for the Cilium Egress Gateway Policies to be synced from k8s to restore and regenerate the endpoints. So on both upgrades and downgrades we will at least have a short connectivity blip.

Did you try to perform an upgrade+downgrade with entries in the map?

jibi replied:
> We currently don't wait for the Cilium Egress Gateway Policies to be synced from k8s to restore and regenerate the endpoints.

I haven't considered that. In this case I think too that with the current state of things there will be connection breakages with upgrades and downgrades (although I haven't tested yet this with entries in the map).

For downgrades I'm not sure there's much we can do given that it's not always possible to map the state of an HA configuration to a non HA one (i.e. it's not clear what we should do when we have multiple gateways for a given policy and downgrade to 1.10 which only supports a single gateway). Moreover egress gateway is still marked as beta, so that may be acceptable?

On the other hand, for upgrades it should be possible to migrate the old map to the new one while the agent has not synced with k8s (but I haven't properly thought about that yet).

pchaigno replied:
> For downgrades I'm not sure there's much we can do given that it's not always possible to map the state of an HA configuration to a non HA one (i.e. it's not clear what we should do when we have multiple gateways for a given policy and downgrade to 1.10 which only supports a single gateway).

Downgrades are most likely to happen before users start using the new egress gateway feature. That is, I'd expect downgrades to happen shortly after upgrades, in case the upgrade has issues. Given that, the egress gateway configuration at the time of the downgrade will likely not be an HA one.

> On the other hand, for upgrades it should be possible to migrate the old map to the new one while the agent has not synced with k8s (but I haven't properly thought about that yet).

Probably safe to just instruct the agent to wait for the k8s sync?

jibi replied:
> Probably safe to just instruct the agent to wait for the k8s sync?

I don't have much context on why we are currently not waiting for k8s to sync but if there's nothing in particular preventing us from doing so then yes, we can change the code to wait for it :+1:

pchaigno replied:
I'd expect it to just be an oversight. It will slow the agent startup a bit, but probably still the right thing to do. Probably best done as a separate PR though, to be able to backport it to v1.10 (that way the downgrade should be fixed as well).

pchaigno commented on bpf/lib/maps.h:
```diff
@@ -186,21 +186,23 @@ struct bpf_elf_map __section_maps ENCRYPT_MAP = {
 };
+
+struct bpf_elf_map __section_maps EGRESS_CT_MAP = {
+	.type		= BPF_MAP_TYPE_LRU_HASH,
```
What's the impact if in-use entries are evicted from the LRU? A new gateway node may be selected and it won't have the SNAT state; will that cause connections to be forcefully terminated? If that's the case, do we need a mechanism to alert the agent when the map is full?

jibi replied:
Correct, if the map is full and a connection gets evicted, it will likely be moved to a different gateway node (at least with the current selection strategy based on `random()`) and it will get reset because of the lack of SNAT state.
At the same time I don't think there's much we can do about it :thinking: having a log in the agent could be useful, but I'm not sure how to properly detect when an element gets evicted (maybe we can just periodically check if the number of elements in the map is close to its size?)

pchaigno replied:
Seems it's also the case for the usual CT map. So let's keep it as is for now :slightly_smiling_face:

pchaigno commented on cilium/cmd/bpf_egress_get.go:
```diff
@@ -6,50 +6,54 @@ package cmd
+		for i := uint32(0); i < val.Size; i++ {
+			gatewayIPs = append(gatewayIPs, val.GatewayIPs[i].String())
+		}
+		fmt.Printf("Gateway IPs: %s\n", strings.Join(gatewayIPs, ", "))
```
Could we put these on newlines for readability?
```suggestion
		fmt.Printf("Gateway IPs:\n")
		for i := uint32(0); i < val.Size; i++ {
			fmt.Printf("\t%s\n", val.GatewayIPs[i].String())
		}
```

pchaigno commented on pkg/egresspolicy/manager.go:
```diff
@@ -287,14 +287,7 @@ func (manager *Manager) updateEgressMap(ips []string, config *Config) error {
-			copy(value.EgressIP[:], config.egressIP)
-
-			err := egressmap.EgressMap.Update(&key, value)
+			err := egressmap.InsertEgressGateway(sip, *dstCIDR, config.egressIP, config.egressIP)
```
This doesn't seem super efficient as we'll make 1 `BPF_MAP_UPDATE_ELEM` call per gateway. We could introduce a `InsertEgressGateway` method to insert all gateways at the same time instead. That said, I'm not sure it really matters since this is only affecting the agent's speed and CPU consumption.

jibi replied:
I thought about this as well. No strong opinions here, but given that I'm not expecting the configuration to change that frequently, I initially found it hard to justify the extra code complexity/method to save a few `BPF_MAP_UPDATE_ELEM` calls (especially given that the maximum number of gateways per policy is currently 64).

pchaigno commented on pkg/maps/egressmap/policy.go:
```diff
@@ -0,0 +1,133 @@
+	*ebpf.Map
+}
+
+// initEgressCtMap initializes the egress policy map.
```
```suggestion
// initEgressPolicyMap initializes the egress policy map.
```

pchaigno commented on cilium/cmd/bpf_egress_list.go:
```diff
@@ -15,20 +17,48 @@ import (
-		bpfEgressList := make(map[string][]string)
-		if err := egressmap.EgressMap.Dump(bpfEgressList); err != nil {
-			Fatalf("error dumping contents of map: %s\n", err)
+		egressmap.InitEgressMaps()
```
IIUC, calling this command when the egress gateway is disabled will cause us to create a dummy map because `InitEgressMaps` ends up calling `OpenOrCreate`. Maybe we should avoid that? I don't think the other `cilium bpf` commands have this side effect.

jibi replied:
Yep makes sense, I'll add a new method to just open the existing pinned map and switch these commands to it :+1:

pchaigno commented on pkg/maps/egressmap/ct.go:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	for _, k := range keysToDelete {
+		log.Infof("Removing egress CT entry: %+v", k)
```
Should this be a debug log instead? It can amount to a lot of logs and we already have higher-level logs anyway.

jibi replied:
On a second thought, it's probably not even worth logging it as a debug log (if necessary it's possible to inspect the CT table with the `bpf egress ct list` command anyway)

pchaigno commented on bpf/lib/egress_gateway.h:
```diff
@@ -0,0 +1,91 @@
+		 *     policy->size < EGRESS_MAX_GATEWAY_NODES"
+		 */
+		index = 0;
+	}
```
Wouldn't the following be nicer w.r.t. complexity?
```c
index %= EGRESS_MAX_GATEWAY_NODES;
```

pchaigno commented on bpf/lib/egress_gateway.h:
```diff
@@ -0,0 +1,91 @@
+
+#include "maps.h"
+
+#if defined(ENABLE_EGRESS_GATEWAY)
```
Nit: I think the preference is on `#ifdef` unless we can't do without `defined`. IIRC, Quentin even translated lots of `#if defined` to `#ifdef` a while back.

pchaigno reviewed (commented):
The code looks great to me! Thanks in particular for the new unit tests and the bit of refactoring :rocket: 

I think my only maybe blocking comment is the potential upgrade concern. Other comments are questions or minor.

The pull request is nevertheless annoying to review because everything is in a single commit. Any chance you could split it into multiple commits? In particular, there are changes that clearly belong to preparatory commits (cleanup, renaming of maps, refactoring) and changes that could be moved to follow up commits (unit tests & CLI changes). That would sensibly ease the review in my opinion because we wouldn't have to navigate all over the place to review each logical piece.

jibi committed:
egressmap: pin connections to egress gateway

This commit introduces a new map, cilium_egress_ct_v4, which is used to
pin connections to a specific egress gateway node.

In this way, after a gateway node is randomly picked for a new
connection, all traffic belonging to that connection will be redirected
to the same gateway.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi committed:
bpf: move egress gateway logic to lib/egress_gateway.h

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi committed:
egressmap: add support for multiple gateway nodes per policy

This commit lays the groundwork (mostly from a datapath perspective) to
implement high availability egress grateway by adding support for
multiple gateway nodes per policy, sharing all the same egress IP.

The full CFP can be found at [1].

The format of the egress policy map has been updated to allow storing
multiple gateways per policy.

Note that these changes assume the router in front of the gateways will
keep track of which ECMP route a connection belongs to (in order to
forward reply traffic back to the correct gateway).
This in practice is not always true, and in a subsequent PR a more
general approach will be introduced to make sure reply traffic will
eventually hit the correct gateway, regarless of which one the router
forwards the traffic to.

Lastly, for the case of an egress policy with a single gateway, this
change should be backward compatible with the current egress gateway
implementation.

[1] https://docs.google.com/document/d/1ojU8bmy6ai8fdClHCVAPN5-SExm5KjLsh_IJ24tY-4M/edit?usp=sharing

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi committed:
egressmap: add unit tests

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

brb commented on bpf/lib/common.h:
```diff
@@ -245,9 +249,16 @@ struct endpoint_info {
+};
+
+struct egress_policy {
+	__u32 size;
```
Could you add a comment explaining this field (w/o reading the code it might be not obvious what this field stands for)?

brb commented on pkg/maps/egressmap/egress.go:
```diff
@@ -4,79 +4,157 @@
+// If the policy does not exist, it will create a new one.
+func InsertEgressGateway(sourceIP net.IP, destCIDR net.IPNet, egressIP, gatewayIP net.IP) error {
+	log.WithFields(logrus.Fields{
+		"source-ip":  sourceIP,
```
Please define logfields (can be done as a follow-up).

brb reviewed (approved):
Nice! Only non-blocking comments.

joestringer commented on Documentation/gettingstarted/egress-gateway.rst:
```diff
@@ -12,6 +12,10 @@ Egress Gateway (beta)
 
+.. note::
+
+   Egress Gateway requires a 5.4 or more recent kernel.
```
Please also add this to `Documentation/operations/system_requirements.rst` in the appropriate section for kernel requirements for particular features.

joestringer commented on bpf/lib/maps.h:
```diff
@@ -186,21 +186,23 @@ struct bpf_elf_map __section_maps ENCRYPT_MAP = {
 	.flags		= BPF_F_NO_PREALLOC,
 };
+
+struct bpf_elf_map __section_maps EGRESS_CT_MAP = {
```
Was there a discussion already about introducing a brand new CT map?

At a high level, the concerns I would have here are around basically:
* This introduces another independent map that could have sizing or LRU issues, which would need to be tracked independently of the other issues. This can make it a bit more complicated to debug when there's a problem due to the map filling up. At least in part maybe this is something that the kernel BPF APIs need to get better at so that we can more easily respond to memory pressure. At the same time, designs like this can exacerbate such issues.
* Additional set of entries to handle garbage collection for, which can be expensive. In theory there should be tied fates with the corresponding entry in the main CT table right? But those tied fates must be implemented somehow, presumably through a tie-in to the main CT GC loop with an additional syscall per GC'd entry. 
* The key is duplicated vs. regular CT map, so we will use at least `EGRESS_CT_MAP_SIZE * sizeof(struct ipv4_ct_tuple)` more memory than if we could integrate this into the existing CT table. Key size here seems to be 14B so this probably puts it in the order of a few extra MB, so maybe not such a big deal.

At the same time, it's worth noting that CT entries represent state that is difficult to reconstruct on-the-fly. The content is generated based on traffic flowing through the system and is highly dynamic, so live-migrating from one map to another can be quite complicated. The result is that if we change the map parameters, it will trigger datapath drops during upgrade. I'm guessing the reason that you avoided modifying the existing CT map is that in order to fit the new entry value there, it would need resizing, thereby triggering this issue. At the same time, it's worth noting that by sizing the 'value' exactly to the size of the data you need for this PR, it also prevents this map from being resized in future without triggering the same sort of problems. Back in the day we had an open question "how big should the CT entry be to future-proof it?" and I think that question becomes relevant again here.

jibi replied:
> Was there a discussion already about introducing a brand new CT map?

Yes, it was discussed in the initial brainstorming sessions I had with Daniel and Christopher, as well as in the CFP.

The reason why we went for a separate CT map approach rather than reusing the existing ones is that the original `CT_MAP_{TCP, ANY}{4,6}` maps are intended to really track connections, their state, stats, rev-nat information etc while this new `EGRESS_CT_MAP` map is supposed to track just which tuple is going to which gateway node, in order to pin it. So calling it a "connection table" might be a bit of a stretch, but given that also the Maglev paper calls it this way, I went for the same name.

At the same time I do agree that this information could be stored in the usual CT maps and that with this approach we are duplicating some data (specifically the `ipv4_tuple` key, across the 2 tables). Another approach would be to extend the existing maps with the `gateway_ip` field, but as you pointed out, that would cause some issues during upgrades/downgrades.

> At the same time, it's worth noting that by sizing the 'value' exactly to the size of the data you need for this PR, it also prevents this map from being resized in future without triggering the same sort of problems.

That's a good point, we could reserve some space in the value for future fields, but at the same time it feels bit arbitrary (i.e. how much space should we reserve?)

> Back in the day we had an open question "how big should the CT entry be to future-proof it?" and I think that question becomes relevant again here.

I just went for the same size we are using for the other CT maps since that size proved to work well in practice, assuming that `egress CT entries <= all CT entries` (while I think in practice it's safe to assume that `egress CT entries << all CT entries`

joestringer replied:
> The reason why we went for a separate CT map approach rather than reusing the existing ones is that the original CT_MAP_{TCP, ANY}{4,6} maps are intended to really track connections, their state, stats, rev-nat information etc while this new EGRESS_CT_MAP map is supposed to track just which tuple is going to which gateway node, in order to pin it. So calling it a "connection table" might be a bit of a stretch, but given that also the Maglev paper calls it this way, I went for the same name.

For what it's worth, we do something similar already with entirely independent entries for the service stickiness, see `CT_SERVICE`. The `tuple->flags` contains values like `CT_INGRESS` / `CT_EGRESS` / `CT_SERVICE` to differentiate the types of info we are storing, and in the service case we are even overloading some of the ct_entry fields already. There may even be a middle ground there where we don't have to resize the ct_entry but we could create new entries with a different CT_* flag to store other metadata about the flow for advanced features. First user would be this feature. It'd take one more entry per flow if you're using this feature, but that's not so bad, it's one less map to deal with.

> That's a good point, we could reserve some space in the value for future fields, but at the same time it feels bit arbitrary (i.e. how much space should we reserve?)

Yeah, this is the difficult part. One way I'm thinking about this is, could these entries be considered a more general "ct extension" where we can hang all sorts of information, and it just happens to be that the first feature that requires more metadata to be associated with a CT entry is this particular gateway stickiness info? You could imagine that over time we accumulate many similar maps like this, but ideally we could just say "alright depending on what you want to do, you may need to attach some additional info, and here's the mechanism to do this".

> > Back in the day we had an open question "how big should the CT entry be to future-proof it?" and I think that question becomes relevant again here.
> 
> I just went for the same size we are using for the other CT maps since that size proved to work well in practice, assuming that egress CT entries <= all CT entries (while I think in practice it's safe to assume that egress CT entries << all CT entries

:+1: Just to be clear, I was asking about the size of the ct_entry itself here (ie value), rather than the overall map size. This is just an extension of the same thoughts / musings about the size of the value above.

> Additional set of entries to handle garbage collection for, which can be expensive. In theory there should be tied fates with the corresponding entry in the main CT table right? But those tied fates must be implemented somehow, presumably through a tie-in to the main CT GC loop with an additional syscall per GC'd entry.

To bring this back up, even if we decide to make no further changes based on the above discussions, I wonder if we need something a bit more for this concern? Like today even with LRU enabled, we enable regular CT garbage collection to ensure we clean up the NAT map if any CT map entries are deleted. I think that the same logic would need to deal with this new map as well? This allows us to clean up entries when pods are created/deleted as well.

jibi replied:
OK, that's a good starting point. I'll play with the idea of storing the gateway related info in the usual CT maps.
I can introduce a new flag (to be used in the `ipv4_ct_tuple.flags` field), for example `TUPLE_F_EGRESS_GATEWAY` and then look for an existing field that can be overloaded. I'm not entirely sure about the implications of this approach for the rest of the datapath/if it's really doable, but worth giving it a try :+1: 

> Like today even with LRU enabled, we enable regular CT garbage collection to ensure we clean up the NAT map if any CT map entries are deleted. I think that the same logic would need to deal with this new map as well?

There's already some logic in place to delete these entries whenever the related policy gets deleted, but I'll double check if I'm missing anything

joestringer commented on bpf/lib/common.h:
```diff
@@ -245,9 +249,23 @@ struct endpoint_info {
+	__be32 gateway_ips[EGRESS_MAX_GATEWAY_NODES];
+};
+
+struct egress_ct {
```
nit, but this can help readability a lot for those unfamiliar with the intent of the code: 
```suggestion
struct egress_policy_entry {
	/* Size is the number of IPs set in the gateway_ips field (i.e. the number of
	 * gateways configured for the policy).
	 */
	__u32 size;
	__be32 egress_ip;
	__be32 gateway_ips[EGRESS_MAX_GATEWAY_NODES];
};

struct egress_ct_entry {
```

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -649,6 +651,16 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+			return nil, nil, fmt.Errorf("cannot determine kernel version: %w", err)
+		}
+
+		if versioncheck.MustCompile("<" + "5.4.0")(kernelVersion) {
```
Is there a way to perform this check without tying it directly to a kernel version? Typically RHEL backports a lot of functionality, and this can mean that Cilium refuses to run this logic on a RHEL kernel when it should really work there.

Ideally this should probe for the actual kernel feature this is relying on from v5.4.

jibi replied:
I think we can probably probe for `HAVE_LARGE_INSN_LIMIT` and use that instead of the kernel version

joestringer reviewed (changes requested):
Nice work. I have one main question below around the introduction of a new CT map below, the rest of my comments are pretty minor nits.

christarazi commented on pkg/logging/logfields/logfields.go:
```diff
@@ -535,4 +535,16 @@ const (
+	EgressIP = "egress-ip"
+
+	// Gateway IP is the gateway IP used in a given egress policy
+	GatewayIP = "gateway-ip"
```
These should be camel case instead given that the above logfields are all camel cased

christarazi commented on pkg/maps/egressmap/policy.go:
```diff
@@ -0,0 +1,143 @@
+// egressCtMap is the internal representation of an egress policy map.
+type egressPolicyMap struct {
+	*ebpf.Map
+}
```
Nit: comment outdated

christarazi commented on pkg/maps/egressmap/egress_privileged_test.go:
```diff
@@ -0,0 +1,166 @@
+	bpf.CheckOrMountFS("")
+	err := bpf.ConfigureResourceLimits()
+	c.Assert(err, IsNil)
+
```
Nit: extra newline

christarazi reviewed (changes requested):
Minor comments from me, given that others have already reviewed from their respective area

jibi committed:
daemon: require kernel >= 5.4 for egress gateway

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

jibi committed:
probes: allow to probe for HAVE_LARGE_INSN_LIMIT support

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>
